### ISSUE TITLE
Strange .cloned() error message mentioning iterators

### ISSUE THREAD

**Issue by enterprisey** [rust-lang/rust#74121]
I'm unsure why the error message for this code mentions iterators; there aren't any, I thought.

```rust
fn main(){
Some(1).cloned();
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5de4c57cd625e8482d579c226cb0c74e))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0599]: no method named `cloned` found for enum `std::option::Option<{integer}>` in the current scope
   --> src/main.rs:2:9
    |
2   |   Some(1).cloned();
    |           ^^^^^^ method not found in `std::option::Option<{integer}>`
    |
    = note: the method `cloned` exists but the following trait bounds were not satisfied:
            `std::option::Option<{integer}>: std::iter::Iterator`
            which is required by `&mut std::option::Option<{integer}>: std::iter::Iterator`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.

```

**Comment by estebank** [comment]
A method `fn cloned(&self)` exists in `Iterator`. The `note` is telling you that method that does exist is not callable on `Option<_>` because `Option<_>` doesn't implement `Iterator`. We can probably reword this to make that clearer.

**Comment by arora-aman** [comment]
Option also has a [`cloned`](https://doc.rust-lang.org/std/option/enum.Option.html#method.cloned) method. Based on the current usecase (not using an iterator) may something along the lines "cloned() defined for `&i32` and not `i32`" along with the iterator lint might be better diagnostically.

**Comment by jazevedo620** [comment]
The solution for the Option case is to use something like:

```rust
fn main(){
  Some(1).as_ref().cloned();
}
```

I encountered the same error, and while I understand the compiler error/message, it sticks out in a toolchain that normally has very helpful compiler error messages. I agree that the help note could be clearer for the case with `Option` (where the intention likely isn't to use `Iterator` at all)

**Comment by ghubertpalo** [comment]
I have got the same error using this test case:

```rust
struct Pika {
    a: Option<u32>,
}

impl Pika {
    pub fn get_a(&self) -> Option<u32> {
        self.a.cloned()
    }
}

fn main() {
    let my_pika = Pika { a: Some(12) };
    println!("{:?}", my_pika.get_a());
}
```
[link to playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=40a2d8a6cae742a1c2995f3ecc703a73)

This grants the following error message:

```txt
Compiling playground v0.0.1 (/playground)
error[[E0599]](https://doc.rust-lang.org/stable/error-index.html#E0599): `Option<u32>` is not an iterator
   --> src/main.rs:7:16
    |
7   |           self.a.cloned()
    |                  ^^^^^^ `Option<u32>` is not an iterator
    |
    = note: the following trait bounds were not satisfied:
            `Option<u32>: Iterator`
            which is required by `&mut Option<u32>: Iterator`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `playground` due to previous error
```

In a more general way, this error message raises as soon as the `cloned()` method is called on a borrow of an Option:

```rust
let a = Some(12);
let b = &a;
let c = b.cloned(); // Iterator error
```
